Mathlib/Data/Nat/Choose/Bounds.lean

```diff
@@ -63,9 +63,20 @@ theorem choose_succ_le_two_pow (n k : ℕ) : (n + 1).choose k ≤ 2 ^ n := by
       · rw [choose_succ_succ', two_pow_succ]
         exact Nat.add_le_add (choose_succ_le_two_pow n k) (choose_succ_le_two_pow n (k + 1))
 
-theorem choose_le_two_pow (n k : ℕ) (p : 0 < n) : n.choose k < 2 ^ n := by
+theorem choose_lt_two_pow (n k : ℕ) (p : 0 < n) : n.choose k < 2 ^ n := by
   refine lt_of_le_of_lt ?_ (Nat.two_pow_pred_lt_two_pow p)
   rw [← Nat.sub_add_cancel p]
   exact choose_succ_le_two_pow (n - 1) k
 
+theorem choose_le_two_pow (n k : ℕ) : n.choose k ≤ 2 ^ n := by
```

Comment:
Here's a golf:
```
theorem choose_le_two_pow (n k : ℕ) : n.choose k ≤ 2 ^ n := by
  obtain (rfl | hk) := eq_zero_or_pos k
  · simp [Nat.one_le_two_pow]
  · obtain (rfl | hn) := eq_zero_or_pos n
    · have : choose 0 k = 0 := by convert Nat.choose_zero_succ (k - 1); omega
      simp [this]
    · exact (Nat.choose_lt_two_pow _ _ hn).le
```

You can use `omega` for both subgoals after the convert,
and if you use the right lemma to case-split, you can avoid some rewrites by dummy hypotheses.

Reply:
It might be worth extracting a lemma that choose 0 k is always <= 1 (either, it's choose 0 0 = 1, or 0 by the above).

Do you think this would be worth it? (If it exists, please use it; `exact?` did not find it, though.)

Reply:
I don't think so and I don't know how useful it would be. I usually just want to bound the whole thing independently of `k` , because it pops up whenever you have to estimate products of derivatives.